File: packages/backend/src/connectors/defillama.ts
import { fetchJson } from "./base.js";

const LLAMA = "https://api.llama.fi";

export interface ProtocolMetrics {
  tvl: number | null;
  tvlChange30dPct: number | null;
  chains: string[];
}

interface DefiLlamaProtocol {
  tvl?: { date: number; totalLiquidityUSD: number }[];
  chains?: string[];
}

export async function fetchProtocol(
  slug: string | null | undefined
): Promise<ProtocolMetrics | null> {
  if (!slug) return null;

  const data = await fetchJson<DefiLlamaProtocol>(`${LLAMA}/protocol/${slug}`);
  if (!data) return null;

  const tvlSeries = data.tvl ?? [];
  const current = tvlSeries.length ? tvlSeries[tvlSeries.length - 1].totalLiquidityUSD : null;

  let change30d: number | null = null;
  if (tvlSeries.length > 30 && current) {
    const prior = tvlSeries[tvlSeries.length - 31].totalLiquidityUSD;
    if (prior) change30d = ((current - prior) / prior) * 100;
  }

  return {
    tvl: current ?? null,
    tvlChange30dPct: change30d,
    chains: data.chains ?? [],
  };
}

File: mobile/lib/connectors/defillama.dart
// Ported from packages/backend/src/connectors/defillama.ts.
import 'http_client.dart';

const _llama = "https://api.llama.fi";

class ProtocolMetrics {
  final double? tvl;
  final double? tvlChange30dPct;
  ProtocolMetrics(this.tvl, this.tvlChange30dPct);
}

Future<ProtocolMetrics?> fetchProtocol(String? slug) async {
  if (slug == null || slug.isEmpty) return null;

  final data = await fetchJson('$_llama/protocol/$slug');
  if (data == null) return null;

  final tvlSeries = (data['tvl'] as List?) ?? [];
  final current = tvlSeries.isNotEmpty ? (tvlSeries.last['totalLiquidityUSD'] as num?)?.toDouble() : null;

  double? change30d;
  if (tvlSeries.length > 30 && current != null) {
    final prior = (tvlSeries[tvlSeries.length - 31]['totalLiquidityUSD'] as num?)?.toDouble();
    if (prior != null && prior != 0) change30d = ((current - prior) / prior) * 100;
  }

  return ProtocolMetrics(current, change30d);
}
